### PR TITLE
Add comment block for insert_shreds_handle_duplicate in blockstore.rs

### PR DESCRIPTION
**Pull request by yhchiang-sol**
#### Summary of Changes
Add code comment for nsert_shreds_handle_duplicate in blockstore.rs

### CODE REVIEW

File: ledger/src/blockstore.rs [hunk 1]
```diff
@@ -770,6 +770,47 @@ impl Blockstore {
         recovered_data_shreds
     }
 
+    /// The main helper function that performs the shred insertion logic
+    /// and updates necessary meta-data.
```
Comment (steviez) [inline ledger/src/blockstore.rs:774]: (very big) nit: "necessary" ==> "corresponding"

**Review by steviez — APPROVED** [review]
Looks good to me - my only comment is a suggestion for your consideration but what you have looks good as-is too!

File: ledger/src/blockstore.rs [hunk 2]
```diff
@@ -770,6 +770,47 @@ impl Blockstore {
...
+    ///
+    /// Arguments:
+    ///  - `shreds`: the shreds to be inserted.
+    ///  - `is_repaired`: a boolean vector aligned with `shreds` where each
+    ///    boolean indicates whether the corresponding shred is repaired or not.
+    ///  - `leader_schedule`: the leader schedule
+    ///  - `is_trusted`: whether the shreds come from a trusted source.
+    ///  - `retransmit_sender`: the sender for transmitting any recovered
+    ///    data shreds.
+    ///  - `handle_duplicate`: a function for handling duplication
```
Comment (carllin) [inline ledger/src/blockstore.rs:808]: nit: a function for handling shreds that have the same slot and index, but different payloads

File: ledger/src/blockstore.rs [hunk 3]
```diff
@@ -770,6 +770,47 @@ impl Blockstore {
...
+    ///   - [`cf::Index`]: stores (slot id, index to the index_working_set_entry)
+    ///     pair to the `cf::Index` column family for each index_working_set_entry
+    ///     which insert did occur in this function call.
+    ///
+    /// Arguments:
+    ///  - `shreds`: the shreds to be inserted.
+    ///  - `is_repaired`: a boolean vector aligned with `shreds` where each
+    ///    boolean indicates whether the corresponding shred is repaired or not.
+    ///  - `leader_schedule`: the leader schedule
+    ///  - `is_trusted`: whether the shreds come from a trusted source.
```
Comment (carllin) [inline ledger/src/blockstore.rs:805]: nit: Passing true skips the duplicate check and shred integrity checks

File: ledger/src/blockstore.rs (hunk 2 above)
Comment (steviez) [reply ledger/src/blockstore.rs:808]: Technically, we don't examine if the payloads are same/different; we just see if we have a payload in for that slot/index combo:
https://github.com/solana-labs/solana/blob/82672b40fd158c09813653586b1937033b194c9a/ledger/src/blockstore.rs#L1236-L1239

https://github.com/solana-labs/solana/blob/82672b40fd158c09813653586b1937033b194c9a/ledger/src/blockstore.rs#L1060-L1064

File: ledger/src/blockstore.rs [hunk 4]
```diff
@@ -770,6 +770,47 @@ impl Blockstore {
         recovered_data_shreds
     }
 
+    /// The main helper function that performs the shred insertion logic
+    /// and updates necessary meta-data.
+    ///
+    /// This function updates the following column families:
+    ///   - [`cf::DeadSlots`]: mark a shred as "dead" if its meta-data indicates
```
Comment (carllin) [inline ledger/src/blockstore.rs:777]: nit: Marks the slot as dead if 
1) We get a new shred `N` marked as the last shred in the slot `S`, but `N.index()` is less than the current `slot_meta.received` for slot `S`
2) The slot is not currently full

This means there's an alternate version of this slot. See `check_insert_data_shred` for more details

File: ledger/src/blockstore.rs (hunk 2 above)
Comment (carllin) [reply ledger/src/blockstore.rs:808]: that's true, the payload examination is done in our `handle_duplicate` implementation. So maybe just "a function for handling shreds that have the same slot and index"

**Pull request merged**